File: evolve.ml
type sensor = (int * int * int * bool)

type tank = (
  string
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * sensor
  * string list
)

let symbols = [
  "~";
  "!";
  "+";
  "-";
  "*";
  "/";
  "%";
  "&";
  "|";
  "^";
  "<<";
  ">>";
  ">";
  ">=";
  "<";
  "<=";
  "=";
  "<>";
  "abs";
  "dup";
  "pop";
  "exch";
  "if";
  "ifelse";
  "mset";
  "mget";
  "{";
  "}";
  "fire-ready?";
  "fire!";
  "set-speed!";
  "get-turret";
  "set-turret!";
  "sensor?";
  "set-led!";
  "random";
]

let make_nucleotide () =
  Random.int ((List.length symbols) + 360 + 360)
  

let rec make_dna = function
  | 0 ->
      []
  | n ->
      (make_nucleotide ()) :: (make_dna (n - 1))
  

let rec program_of_dna = function
  | [] ->
      []
  | i :: tl ->
      let s =
        try
          List.nth symbols i
        with
          | Failure _
          | Invalid_argument _ ->
              string_of_int (i - (List.length symbols) - 360)
      in
        s :: (program_of_dna tl)

let tank_of_dna d =
  let color_of_dna = function
    | r :: g :: b :: tl ->
        (tl, (Format.sprintf "%02x%02x%02x"
                ((abs r) mod 256)
                ((abs g) mod 256)
                ((abs b) mod 256)))
    | tl ->
        (tl, "cccccc")
  in
  let sensor_of_dna = function
    | a :: b :: c :: d :: tl ->
        (tl, (a, b, c, (d != 0)))
    | tl ->
        (tl, (0, 0, 0, false))
  in
  let d, color = color_of_dna d in
  let d, s0 = sensor_of_dna d in
  let d, s1 = sensor_of_dna d in
  let d, s2 = sensor_of_dna d in
  let d, s3 = sensor_of_dna d in
  let d, s4 = sensor_of_dna d in
  let d, s5 = sensor_of_dna d in
  let d, s6 = sensor_of_dna d in
  let d, s7 = sensor_of_dna d in
  let d, s8 = sensor_of_dna d in
  let d, s9 = sensor_of_dna d in
  let p = program_of_dna d in
    (color, s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, p)


let print_tank t =
  let print_sensor (a, b, c, d) =
    Format.printf "Sensor: %d %d %d %d\n" a b c (if d then 1 else 0)
  in
    match t with
      | (color, s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, p) ->
          Format.printf "Color: #%s\n" color;
          print_sensor s0;
          print_sensor s1;
          print_sensor s2;
          print_sensor s3;
          print_sensor s4;
          print_sensor s5;
          print_sensor s6;
          print_sensor s7;
          print_sensor s8;
          print_sensor s9;
          print_endline (String.concat " " p)

let write_tank i t =
  let fnprint fn str =
    let filename = Format.sprintf "tank.%d/%s" i fn in
    let chan = open_out filename in
      output_string chan str;
      close_out chan
  in
  let print_sensor n (a, b, c, d) =
    fnprint
      (Format.sprintf "sensor%d" n)
      (Format.sprintf "Sensor: %d %d %d %d\n" a b c (if d then 1 else 0))
  in
  match t with
    | (color, s0, s1, s2, s3, s4, s5, s6, s7, s8, s9, p) ->
        fnprint "color" color;
        print_sensor 0 s0;
        print_sensor 1 s1;
        print_sensor 2 s2;
        print_sensor 3 s3;
        print_sensor 4 s4;
        print_sensor 5 s5;
        print_sensor 6 s6;
        print_sensor 7 s7;
        print_sensor 8 s8;
        print_sensor 9 s9;
        fnprint "program" (String.concat " " p)
            
let _ = Random.self_init ()

let _ =
  let dnas =
    let t = Array.make 20 500 in
      Array.map make_dna t
  in
  let tanks = Array.map tank_of_dna dnas in
    Array.mapi write_tank tanks

File: evolve.go
package main

import (
	"fmt"
	"math/rand"
	"strconv"
	"time"
	"strings"
	"io/ioutil"
	"encoding/json"
)

type sensor struct {
	srange int
	angle int
	width int
	turret bool
}

type Tank struct {
	dna []int
	color string
	sensors [10]sensor
	program []string
}

var symbols []string = []string{
	"~",
  "!",
  "+",
  "-",
  "*",
  "/",
  "%",
  "&",
  "|",
  "^",
  "<<",
  ">>",
  ">",
  ">=",
  "<",
  "<=",
  "=",
  "<>",
  "abs",
  "dup",
  "pop",
  "exch",
  "if",
  "ifelse",
  "mset",
  "mget",
  "{",
  "}",
  "fire-ready?",
  "fire!",
  "set-speed!",
  "get-turret",
  "set-turret!",
  "sensor?",
  "set-led!",
  "random",
}

func make_nucleotide() int {
	switch rand.Intn(3) {
	case 0:
	  return rand.Intn(len(symbols))
	case 1:
	  return len(symbols) + rand.Intn(720)
	 case 2:
		return len(symbols) + 720 + rand.Intn(100)
	}
	return 0
}

func make_dna(size int) []int {
	ret := make([]int, size)
	for i := 0; i < size; i += 1 {
		ret[i] = make_nucleotide()
	}
	return ret
}

func color_of_dna(dna []int) (string, []int) {
	if len(dna) >= 3 {
		r := dna[0] % 256
		g := dna[1] % 256
		b := dna[2] % 256
		color := fmt.Sprintf("%02x%02x%02x", r, g, b)
		return color, dna[3:]
	} else {
		return "cccccc", dna
	}
}

func sensor_of_dna(dna []int) (sensor, []int) {
	if len(dna) >= 4 {
		return sensor{dna[0], dna[1], dna[2], dna[3] % 2 == 0}, dna[4:]
	} else {
		return sensor{0, 0, 0, false}, dna
	}
}

func program_of_dna(dna []int) []string {
	ret := make([]string, 0, len(dna))
	
	stacks := []int{}
	for i := 0; i < len(dna); i += 1 {
		// If we are in a stack, decrement the size
		if len(stacks) > 0 {
			stacks[0] -= 1
			// Are we at the end of the stack? Output the symbol and pop the stack off.
			if stacks[0] <= 0 {
				ret = append(ret, "}")
				stacks = stacks[1:]
			}
		}
		
		// Consider the next nucleotide
		n := dna[i]

		// Symbol?
		if n < len(symbols) {
			ret = append(ret, symbols[n])
			continue
		}
		n -= len(symbols)
		
		// Number?
		if n < 720 {
			ret = append(ret, strconv.Itoa(n - 360))
			continue
		}
		n -= 720
		
		// Stack.
		// XXX: This doesn't seem right, we shouldn't see a ton of "} } }" at the end
		ret = append(ret, "{")
		stacks = append(stacks, (n + 1) % (len(dna) - i))
	}
	
	for len(stacks) > 0 {
		ret = append(ret, "}")
		stacks = stacks[1:]
	}
	
	return ret
}

func tank_of_dna(dna []int) Tank {
	ret := Tank{dna: dna}
	ret.color, dna = color_of_dna(dna)
	for i := 0; i < len(ret.sensors); i += 1 {
		ret.sensors[i], dna = sensor_of_dna(dna)
	}
	ret.program = program_of_dna(dna)
	
	return ret
}

func twrite(num int, fn string, contents string) {
	path := fmt.Sprintf("tank.%d/%s", num, fn)
	ioutil.WriteFile(path, []byte(contents), 0644)
}

func (t Tank) Write(num int, dir string) {
	twrite(num, "color", t.color)
	for i := 0; i < len(t.sensors); i += 1 {
		turret := 0
		if t.sensors[i].turret {
			turret = 1
		}
		twrite(
			num,
			fmt.Sprintf("sensor%d", i),
			fmt.Sprintf("%d %d %d %d", t.sensors[i].srange, t.sensors[i].angle, t.sensors[i].width, turret),
		)
	}
	twrite(num, "program", strings.Join(t.program, " "))
	
	s, _ := json.Marshal(t.dna)
	twrite(num, "dna", string(s))
}

func init() {
	rand.Seed(time.Now().UTC().UnixNano())
}

func main() {
	d := make_dna(200)
	t := tank_of_dna(d)
	t.Write(0, ".")
}